---
title: "Figure S1: Spatial Mouse Atlas - Seqfish - Mouse Embryo"
author:
- name: Elyas Heidari
  affiliation: 
  - &IMLS Institute for Molecular Life Sciences, University of Zurich, Switzerland
  - Swiss Institute of Bioinformatics (SIB), University of Zurich, Switzerland
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output:
  workflowr::wflow_html:
    code_folding: hide
    toc: true
    toc_float: true
    number_sections: false
---

# Setup / definitions


```{r setup_knitr, include=FALSE}
library('BiocStyle')
set.seed(1996)
options(bitmapType='cairo')
knitr::opts_chunk$set(autodep=TRUE, cache=FALSE, dev='pdf', cache.lazy = FALSE)
# knitr::opts_chunk$set( autodep=TRUE, cache=TRUE, cache.lazy=FALSE, dev='png' )
knitr::opts_knit$set( root.dir='..' )
# wflow_build(files='analysis/Fig1_dummy.rmd', view=F, verbose=T, delete_cache=T)
```

```{r setup_libs, collapse=FALSE, message=FALSE, warning=FALSE, cache=FALSE}
```

## Libraries

```{r libs, message=FALSE, warning=FALSE, cache=FALSE}
# library('pika')
library('tidyverse')
library('igraph')
library('data.table')
library('SingleCellExperiment')
library('magrittr')
library('BiocParallel')
library('patchwork')
library('limma')
library('cccd')
library('philentropy')
library('cowplot')
library('zellkonverter')
library('stringr')
library('lisi')
library('ggpubr')
library('Rtsne')
# library('anndata')
```


## Helper functions
```{r setup_helpers, message=FALSE, cache=FALSE}
source('code/utils/utils.R')
source('code/utils/SMA_functions.R')
```

## Directories
```{r setup_input, message=FALSE, cache=FALSE}
tag       = 'seqfish_mouse_embryo' 
ref_tag   = 'embryo2_2'
out_dir   = file.path('output', tag)
obj_dir   = 'int_data/seqfish_mouse_embryo/embryo2_2'
dir.create(obj_dir)
obj_list  = list()
fig_dir   = 'figures/seqfish_mouse_embryo/embryo2_2'
dir.create(fig_dir)
fig_list = list()

methods     = list.files(out_dir)
names(methods) = methods
preds_f   = methods %>%
  map(~file.path(out_dir, .x, paste0(paste('preds', ref_tag, 'query_data', sep='_'), '.h5ad')))
imp_f = file.path('int_data/seqfish_mouse_embryo', 'gene_dt_embryo2_2.txt')
col_dt_f = file.path('int_data/seqfish_mouse_embryo', 'col_dt_embryo2_2.txt')

```





## Params
```{r setup_outputs, message=FALSE, cache=FALSE}
n_markers = 50
overwrite = FALSE
```

# Load inputs
```{r load_inputs, message=FALSE, cache=FALSE}
sce = preds_f[['sagenet']] %>% readH5AD 
cell_idx = which(sce$embryo != 'scRNAseq')
sce %<>% .[, cell_idx]
meta_dt = colData(sce) %>% as.data.table
if(overwrite | 
  !file.exists(file.path(obj_dir, 'map_2d.Rds')) | 
  !file.exists(file.path(obj_dir, 'meta_dt.Rds'))
){
  # Gene importanc

  preds_list = list()
  preds_list[['sagenet']] = anndata::read_h5ad(preds_f[['sagenet']])$obsm$dist_map 
  preds_list[['sagenet']]  %<>% as('sparseMatrix') %>% .[cell_idx, cell_idx]
  ## Novosparc predictions


gc()
}

# imp = imp_f %>% fread %>% setnames('V1', 'GENE') %>%
#   .[order(embryo2_2_entropy)] 
# markers = imp$GENE[1:n_markers]
col_dt  = col_dt_f %>%  fread 
cell_type_map = data.table(celltype=unique(col_dt$cell_type)) 
cell_type_map %>% 
  .[, code:=sample(c('A', 'B', 'C', 'D', 'E', 'F'), .N, replace=TRUE)] %>%
  setkey(celltype)
```

# Processing / calculations
## Distance matrices and embeddings
```{r calc_dists, message=FALSE, cache=FALSE}
if(overwrite | 
  !file.exists(file.path(obj_dir, 'map_2d.Rds')) | 
  !file.exists(file.path(obj_dir, 'meta_dt.Rds'))
){
  d = list()
  d_mtx = list()
  map_2d = list()

  # sagenet distances
  d[['sagenet']] = preds_list[['sagenet']] %>% as.dist
  d_mtx[['sagenet']] = d[['sagenet']] %>% as.matrix 
  rownames(d_mtx[['sagenet']]) = colnames(d_mtx[['sagenet']]) = colnames(sce)
  map_2d[['sagenet']] = Rtsne(d_mtx[['sagenet']], is_distance=T, perplexity=50)$Y
  rownames(map_2d[['sagenet']]) = colnames(sce)
  meta_dt[, `:=`(sagenet_1 = ..map_2d[['sagenet']][cell_id, 1], 
                sagenet_2 = ..map_2d[['sagenet']][cell_id, 2])]

  # obj_list[['d_mtx']] = append(d_mtx, d_mtx_true)
  obj_list[['map_2d']] = map_2d
  obj_list[['meta_dt']] = meta_dt
  gc()
}else{
  map_2d = readRDS(file.path(obj_dir, 'map_2d.Rds'))
  meta_dt = readRDS(file.path(obj_dir, 'meta_dt.Rds'))
}
```


## A: True regional labels + cell types (ref)
```{r A, fig.height=2.5, fig.width=15, message=FALSE, cache=FALSE}
cols = c('A' = '#FBBE92', 'B' = '#7BAFDE', 'C' = '#B2DF8A', 'D' = '#8870ad', 'E' = '#E1C239', 'F'= '#c19f70')
names(cols) = NULL
p = plot_spatial(col_dt[order(cell_type), .(-x, y)], labels=NA, label_cols=cols, hide_legend = TRUE, title = 'spatial reference') 
p1 = plot_spatial(col_dt[, .(-x, y)], labels=factor(col_dt$leiden_0.01), hide_legend = TRUE, title='partitioning 1')
p2 = plot_spatial(col_dt[, .(-x, y)], labels=factor(col_dt$leiden_0.05), hide_legend = TRUE, title='partitioning 2')
p3 = plot_spatial(col_dt[, .(-x, y)], labels=factor(col_dt$leiden_0.1), hide_legend = TRUE, title='partitioning 3')
p4 = plot_spatial(col_dt[, .(-x, y)], labels=factor(col_dt$leiden_0.5), hide_legend = TRUE, title='partitioning 4')
p5 = plot_spatial(col_dt[, .(-x, y)], labels=factor(col_dt$leiden_1), hide_legend = TRUE, title='partitioning 5')

fig_list[['A']] = p + p1 + p2 + p3 + p4 + p5 + plot_layout(nrow=1) 
fig_list[['A']]
# plist
```

## B: UMAP plosts of expression space and the reconstructed space
```{r B, fig.height=5, fig.width=10, message=FALSE, cache=FALSE, eval=TRUE}
meta_dt[, ent :=  5 - (ent_embryo2_2_leiden_1 + ent_embryo2_2_leiden_0.1 + ent_embryo2_2_leiden_0.01 + ent_embryo2_2_leiden_0.5 + ent_embryo2_2_leiden_0.05) ]
meta_dt$ent = (meta_dt$ent/(max(meta_dt$ent)))^2
meta_dt %<>% .[, code := ..cell_type_map[as.character(cell_type)]$code] %>%
  .[!is.na(code)] %>%
  .[order(code)]

sagenet_p = plot_2d( 
    dim_df = meta_dt[, .(sagenet_1, sagenet_2)],
    labels = NA,
    label_cols = cols,
    hide_legend = TRUE,
    title = "SageNet's reconstructed space",
    sz = 3,
    alpha = 1,
) + labs(x = '', y = '')

legend1 = (plot_2d( 
    dim_df = meta_dt[, .(sagenet_1, sagenet_2)],
    labels = meta_dt$code,
    label_cols = cols,
    hide_legend = FALSE,
    # title = 'SageNet',
    sz = 5,
    # alpha = meta_dt$ent_plot,
) + labs(x = '', y = '', color = 'cell type')+ theme(text = element_text(size = 15))   ) %>% get_legend() %>%
  as_ggplot() 



fig_list[['B']] = sagenet_p + legend1 + plot_layout(nrow=1) 
fig_list[['B']]


sagenet_p = plot_2d_cont( 
    dim_df = meta_dt[, .(sagenet_1, sagenet_2)],
    labels = meta_dt$ent,
    # label_cols = cols,
    hide_legend = TRUE,
    title = "mapping confidence",
    sz = 3
) + labs(x = '', y = '') +
  # theme_minimal() +
  theme(axis.ticks.x=element_blank(),
    axis.ticks.y=element_blank(), 
    panel.grid.major = element_blank(), 
    panel.grid.minor = element_blank(),
    plot.title = element_text(hjust = 0.5))

legend1 = (plot_2d_cont( 
    dim_df = meta_dt[, .(sagenet_1, sagenet_2)],
    labels = meta_dt$ent,
    # label_cols = cols,
    hide_legend = FALSE,
    title = "SageNet's reconstructed space",
    sz = 3
) + labs(x = '', y = '', color = 'confidence score')+ theme(text = element_text(size = 15)) + scale_fill_gradientn(colors = viridis(10), labels = c("low", "high"))  ) %>% get_legend() %>%
  as_ggplot() 





fig_list[['C']] = sagenet_p + legend1 + plot_layout(nrow=1) 
fig_list[['C']]
```

```{r C, fig.height=4, fig.width=2, message=FALSE, cache=FALSE, eval=FALSE}
meta_dt[, ent :=  5 - (ent_embryo2_2_leiden_1 + ent_embryo2_2_leiden_0.1 + ent_embryo2_2_leiden_0.01 + ent_embryo2_2_leiden_0.5 + ent_embryo2_2_leiden_0.05) ]
meta_dt$ent = (meta_dt$ent/(max(meta_dt$ent)))^2
# tab_dt = col_dt[, .N, by = 'cell_type'] %>% .[order(N), ]
tab_dt_query = copy(meta_dt[, .(ent=median(ent), .N, dataset='query'), by = 'code']) %>% .[order(ent), ]
celltype_ord = unique(tab_dt_query$code)

data_summary <- function(x) {
   m <- median(x)
   ymin <- m-sd(x)
   ymax <- m+sd(x)
   return(c(y=m,ymin=ymin,ymax=ymax))
}

p1 = meta_dt %>%
  ggplot +
  aes(x = factor(code, ordered=TRUE, levels= celltype_ord), y = ent, color=code) +
  # geom_violin() +
  # + geom_jitter(height = 0, width = 0.1) +
  theme_bw() +
  scale_color_manual(values=cols, na.value='#e6e6e6', drop=TRUE) +
  labs(y='confidence score', x= '', title='mapping confidence') +
  stat_summary(fun.data=data_summary, size=1) +
  coord_flip() + 
  theme(plot.title = element_text(hjust = 0.5)) +
  theme(legend.position='none')

p1
```




# Save objects & plots
```{r save, fig.height=12, fig.width=12, message=FALSE, cache=FALSE}
names(obj_list) %>% map(~saveRDS(obj_list[[.x]], file.path(obj_dir, paste0(.x, '.Rds'))))
names(fig_list) %>% map(~saveRDS(fig_list[[.x]], file.path(fig_dir, paste0(.x, '.Rds'))))
```
