---
title: "Developing human heart - Spatial Transcriptomics, Figure 3A-D"
author:
- name: Elyas Heidari
  affiliation: 
  - &IMLS Institute for Molecular Life Sciences, University of Zurich, Switzerland
  - Swiss Institute of Bioinformatics (SIB), University of Zurich, Switzerland
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output:
  workflowr::wflow_html:
    code_folding: hide
    toc: true
    toc_float: true
    number_sections: false
---

# Setup / definitions


```{r setup_knitr, include=FALSE}
library('BiocStyle')
set.seed(1996)
options(bitmapType='cairo')
knitr::opts_chunk$set(autodep=TRUE, cache=FALSE, dev='pdf', cache.lazy = FALSE)
# knitr::opts_chunk$set( autodep=TRUE, cache=TRUE, cache.lazy=FALSE, dev='png' )
knitr::opts_knit$set( root.dir='..' )
# wflow_build(files='analysis/Fig4A_D.rmd', view=F, verbose=T, delete_cache=T)
```

```{r setup_libs, collapse=FALSE, message=FALSE, warning=FALSE, cache=FALSE}
```

## Libraries

```{r libs, message=FALSE, warning=FALSE, cache=FALSE}
# library('pika')
library('tidyverse')
library('igraph')
library('data.table')
library('SingleCellExperiment')
library('magrittr')
library('BiocParallel')
library('patchwork')
library('limma')
library('cccd')
library('philentropy')
library('cowplot')
library('zellkonverter')
library('stringr')
library('lisi')
library('ggpubr')
library('Rtsne')
# library('anndata')
```


## Helper functions
```{r setup_helpers, message=FALSE, cache=FALSE}
source('code/utils/utils.R')
source('code/utils/SMA_functions.R')
```

## Directories
```{r setup_input, message=FALSE, cache=FALSE}
tag       = 'ST_human_heart' 
ref_tag   = 'ST'
data_dir  = file.path('data_tidy', tag)
out_dir   = file.path('output', tag)
obj_dir   = 'int_data/ST_human_heart'
dir.create(obj_dir)
obj_list  = list()
fig_dir   = 'figures/ST_human_heart'
dir.create(fig_dir)
fig_list = list()
data_f    = list.files(data_dir)
exp_names = str_replace(data_f[grep('h5ad$', data_f)], '.h5ad', '')
exp_names = setdiff(exp_names[!grepl('_lo', exp_names)], c('ST', 'scRNAseq', 'query'))
names(exp_names) = exp_names
sce_f     = exp_names %>% 
  map(~file.path(data_dir, paste0(.x, '.h5ad')))
methods     = list.files(out_dir)
names(methods) = methods
preds_f   = methods %>% 
  map(function(y) exp_names %>% 
    map(~file.path(out_dir, y, paste0(paste('preds', paste0(.x, '_lo'), .x, sep='_'), '.h5ad')))
  )

imp_f = file.path(obj_dir, 'imp_ST.txt')
col_dt_f = file.path(obj_dir, 'col_dt_ST.txt')
```





## Params
```{r setup_outputs, message=FALSE, cache=FALSE}
n_markers = 10
overwrite = FALSE
```

# Load inputs
```{r load_inputs, message=FALSE, cache=FALSE}
imp = imp_f %>% fread %>% setnames('V1', 'GENE') %>% .[-1,] %>%
    .[!grepl('^MT', GENE)]
genes = unlist(imp[,1])
imp %<>% .[,-1] %>% as.matrix
rownames(imp) = genes
imp_dt = apply(imp, 2, function(x) rownames(imp)[order(-x)]) %>%
    as.data.table 
imp_dt %<>% .[1:n_markers,]
markers = imp_dt %>% unlist %>% unique
sce_list = sce_f %>% map(readH5AD) 
s_sce = sce_list %>% purrr::reduce(cbind)
meta_dt = colData(s_sce) %>% as.data.table
rm(sce_list)

if(overwrite | 
  !file.exists(file.path(obj_dir, 'map_2d.Rds')) | 
  !file.exists(file.path(obj_dir, 'meta_dt.Rds'))
){
  # Gene importanc

  preds_list = list()
  preds_list[['sagenet']] = preds_f[['sagenet']] %>% map(~anndata::read_h5ad(.x)$obsm$dist_map %>% as('sparseMatrix')) 
  ## Novosparc predictions
  preds_list[['novosparc']] = preds_f[['novosparc']] %>% map(read_preds_novosparc) 
  gc()
  # Tangram predictions
  preds_list[['tangram_cells']] = preds_f[['tangram_cells']] %>% map(read_preds_tangram)
  gc()


gc()
}

col_dt  = col_dt_f %>%  fread %>%
  .[, class__ := factor(..class__code[class__], ordered=TRUE, levels=class__ord)]

grid_code = c(
  "6" = 'top - left',
  "7" = 'top - middle',
  "8" = 'top - right',
  "3" = 'middle - left',
  "4" = 'middle - middle',
  "5" = 'middle - right',
  "0" = 'bottom - left',
  "1" = 'bottom - middle',
  "2" = 'bottom - right'
)
```

# Processing / calculations
## Distance matrices and embeddings
```{r calc_dists, message=FALSE, cache=FALSE}
if(overwrite | 
  !file.exists(file.path(obj_dir, 'map_2d.Rds')) | 
  !file.exists(file.path(obj_dir, 'meta_dt.Rds'))
){
  d      = list()
  d_mtx  = list()
  map_2d = list()

  map_2d[['tangram_cells']] = preds_list[['tangram_cells']]
  map_2d[['true']]          = as.matrix(meta_dt[, .(x,y)])
  rownames(map_2d[['true']]) = meta_dt$cell_id
  map_2d[['novosparc']]     = preds_list[['novosparc']] 


  # sagenet distances
  d[['sagenet']] = preds_list[['sagenet']] %>% map(as.dist)
  d_mtx[['sagenet']] = d[['sagenet']] %>% map(as.matrix)
  d[['tangram_cells']] = preds_list[['tangram_cells']] %>% map(dist)
  d_mtx[['tangram_cells']] = d[['tangram_cells']] %>% map(as.matrix)
  d[['novosparc']] = preds_list[['novosparc']] %>% map(dist)
  d_mtx[['novosparc']] = d[['novosparc']] %>% map(as.matrix)

  for(i in exp_names){
    rownames(d_mtx[['sagenet']][[i]]) = colnames(d_mtx[['sagenet']][[i]]) = colnames(s_sce[, s_sce$sample == i])
    map_2d[['sagenet']][[i]] = Rtsne(d_mtx[['sagenet']][[i]], is_distance=T)$Y
    rownames(map_2d[['sagenet']][[i]]) = colnames(s_sce[, s_sce$sample == i])

    d[['exprs']][[i]] = dist(t(assays(s_sce[, s_sce$sample == i])[[1]])) %>% as.dist
    d_mtx[['exprs']][[i]] = d[['exprs']][[i]] %>% as.matrix 
    rownames(d_mtx[['exprs']][[i]]) = colnames(d_mtx[['exprs']][[i]]) = colnames(s_sce[, s_sce$sample == i])
    map_2d[['exprs']][[i]] = Rtsne(d_mtx[['exprs']][[i]], is_distance=T)$Y
    rownames(map_2d[['exprs']][[i]]) = colnames(s_sce[, s_sce$sample == i])

    d[['sagenet_markers']][[i]] = dist(t(assays(s_sce[markers, s_sce$sample == i])[[1]])) %>% as.dist
    d_mtx[['sagenet_markers']][[i]] = d[['sagenet_markers']][[i]] %>% as.matrix 
    rownames(d_mtx[['sagenet_markers']][[i]]) = colnames(d_mtx[['sagenet_markers']][[i]]) = colnames(s_sce[, s_sce$sample == i])
    map_2d[['sagenet_markers']][[i]] = Rtsne(d_mtx[['sagenet_markers']][[i]], is_distance=T)$Y
    rownames(map_2d[['sagenet_markers']][[i]]) = colnames(s_sce[, s_sce$sample == i])

    rownames(d_mtx[['tangram_cells']][[i]]) = colnames(d_mtx[['tangram_cells']][[i]]) = colnames(s_sce[, s_sce$sample == i])
    rownames(map_2d[['tangram_cells']][[i]]) = colnames(s_sce[, s_sce$sample == i])


    rownames(d_mtx[['novosparc']][[i]]) = colnames(d_mtx[['novosparc']][[i]]) = colnames(s_sce[, s_sce$sample == i])
    rownames(map_2d[['novosparc']][[i]]) = colnames(s_sce[, s_sce$sample == i])

  }

  # true physical distances
  d_true = dist(meta_dt[, .(x,y)])
  d_mtx_true = d_true %>% as.matrix
  rownames(d_mtx_true) = colnames(d_mtx_true) = meta_dt$cell_id

  # obj_list[['d_mtx']] = append(d_mtx, d_mtx_true)
  obj_list[['map_2d']] = map_2d
  obj_list[['meta_dt']] = meta_dt
  gc()
}else{
  map_2d = readRDS(file.path(obj_dir, 'map_2d.Rds'))
  meta_dt = readRDS(file.path(obj_dir, 'meta_dt.Rds'))
}
```


## Correlation between the true distance matrix and the prediction distance matrix
```{r calc_cor, message=FALSE, cache=FALSE}
if(overwrite | 
  !file.exists(file.path(obj_dir, 'cor_list.Rds'))
){
  cor_list    = list()
  for(meth in setdiff(names(map_2d), 'true')){

  # for(meth in 'sagenet_markers'){
    cor_list[[meth]]    = c()
    for(e in levels(factor(meta_dt$sample))){
      cells = meta_dt[sample == e]$cell_id
      cor_list[[meth]][e] = cor(c(d_mtx_true[cells, cells]), c(d_mtx[[meth]][[e]]), method='spearman')
    }
  }

  obj_list[['cor_list']] = cor_list
  gc()
}else{
  cor_list = readRDS(file.path(obj_dir, 'cor_list.Rds'))
}

cor_dt = names(cor_list) %>%
  map(~data.table(method=.x, sample=names(cor_list[[.x]]), corr=cor_list[[.x]])) %>%
  purrr::reduce(rbind)
```

## Cell2Cell afinity scores
```{r calc_c2c, message=FALSE, cache=FALSE}
if(overwrite | 
  !file.exists(file.path(obj_dir, 'ccc_list.Rds'))
){
  ccc_list    = list()
  # True physical space
  ccc_list[['true']] = list()
  for(e in levels(factor(meta_dt$sample))){
    
    cells = meta_dt[sample == e]$cell_id
    g_out    = map_2d[['true']] %>% 
      .[cells, ] %>% 
      get_delaunay(plot=FALSE) %>%
      .$graph
    ccc_list[['true']][[e]] = cellCellContact(
        sce = s_sce[, cells],
        group = 'class__',
        graph = g_out,
        nperm = 500,
        plot = FALSE,
        cellID='cell_id')
  }

  # Predicted spaces

  for(meth in setdiff(names(map_2d), 'true')){
  # for(meth in 'sagenet_markers'){
    ccc_list[[meth]] = list()
    for(e in levels(factor(meta_dt$sample))){
      cells = meta_dt[sample == e]$cell_id
      rownames(map_2d[[meth]][[e]]) = cells
      g_out    = map_2d[[meth]][[e]] %>% 
        get_delaunay(plot=FALSE) %>%
        .$graph
      ccc_list[[meth]][[e]] = cellCellContact(
          sce    = s_sce[, cells],
          group  = 'class__',
          graph  = g_out,
          nperm  = 500,
          plot   = FALSE,
          cellID ='cell_id')
    }
  }

  obj_list[['ccc_list']] = ccc_list
}else{
  ccc_list = readRDS(file.path(obj_dir, 'ccc_list.Rds'))
}
ccc_dist    = list()
for(meth in setdiff(names(map_2d), 'true')){
  ccc_dist[[meth]] = c()
  for(e in levels(factor(meta_dt$sample))){
    m1 = ccc_list[[meth]][[e]]$pmat 
    m2 = ccc_list[['true']][[e]]$pmat
    ccc_dist[[meth]][e] = norm(m1 - m2 , '2')
  }
}

# ccc_plot_list = names(ccc_list) %>%
#   map(~ccc_list[[.x]][3] %>% map(function(y) cellCellContactHeatmapTriangle(y, celltype_colours, title=.x)))

# Scores' data table 
ccc_dt = names(ccc_dist) %>%
  map(~names(ccc_dist[[.x]]) %>% map(function(y) 
      data.table(method=.x, sample=y, corr=ccc_dist[[.x]][y])
    ) %>% purrr::reduce(rbind) 
  ) %>%
  purrr::reduce(rbind) 


```



## Distribution of the lisi scores
```{r calc_lisi, message=FALSE, cache=FALSE}
preds_list = list()
if(overwrite | 
  !file.exists(file.path(obj_dir, 'lisi_vals.Rds'))
){
  preds_list[['sagenet']] = map_2d[['sagenet']]
  preds_list[['exprs']] = map_2d[['exprs']]
  preds_list[['sagenet_markers']] = map_2d[['sagenet_markers']]
  preds_list[['true']] = names(map_2d[['sagenet']]) %>% map(~meta_dt[sample==.x, .(x,y)]) 
  names(preds_list[['true']]) = names(map_2d[['sagenet']])
  preds_list[['novosparc']] = map_2d[['novosparc']]
  preds_list[['tangram_cells']] = map_2d[['tangram_cells']]
  # lisi_list   = list()
  lisi_vals = names(preds_list) %>%
    map(function(y) names(preds_list[[y]]) %>%
      map(~compute_lisi(preds_list[[y]][[.x]], meta_dt[sample==.x], c("class__")) %>%
        as.data.table %>% 
        .[, `:=`(cell_id = ..meta_dt[sample==.x]$cell_id, method = y, sample = .x)]) 
    ) %>% map(~.x %>% purrr::reduce(rbind)) %>% purrr::reduce(rbind)
    
  obj_list[['lisi_vals']] = lisi_vals
}else{
  lisi_vals = readRDS(file.path(obj_dir, 'lisi_vals.Rds'))
}

lisi_vals = dcast(lisi_vals, cell_id ~ method, value.var='LISI' ) %>%
  .[, (setdiff(names(.), 'cell_id')) := lapply(.SD, function(x) x/.SD$true), .SDcols = setdiff(names(.), 'cell_id')] %>%
  melt(variable.name='method', value.name='LISI') %>%
  .[method != 'true']
methods = setdiff(unique(lisi_vals$method), 'true')
names(methods) = methods
lisi_vals %<>% setkey(cell_id)
lisi_vals = lisi_vals[order(cell_id)] %>%
  .[meta_dt[, .(cell_id, class__, sample)]]
```

## 4A: True regional labels + grids + reconstruction (ref)
```{r 4A, fig.height=24, fig.width=10, message=FALSE, cache=FALSE}
meta_dt[, class__ := factor(..class__code[class__], ordered=TRUE, levels=class__ord)]
grid2_cols = .palette1[1:length(unique(col_dt$grid_2))]
names(grid2_cols) = as.factor(as.character(unique(col_dt$grid_2)))
grid3_cols = .palette1[1:length(unique(col_dt$grid_3))]
names(grid3_cols) = as.factor(as.character(unique(col_dt$grid_3)))
grid4_cols = .palette1[1:length(unique(col_dt$grid_4))]
names(grid4_cols) = as.factor(as.character(unique(col_dt$grid_4)))
for(i in 1:length(setdiff(exp_names, 'query'))){
  c_dt   = col_dt[sample == exp_names[i]] %>% .[order(as.character(class__))]
  if(i == 1){
    p = plot_spatial(c_dt[, .(x, y)], labels=factor(as.character(c_dt$class__)), hide_legend = TRUE, title='region', label_cols=class__cols, sz=2, label_title='region') + coord_fixed()  + labs(y=exp_names[i]) + theme(text = element_text(size = 20))

    legend_region = (plot_spatial(c_dt[, .(x, y)], labels=factor(as.character(c_dt$class__)), hide_legend = FALSE, title='region', label_cols=class__cols, sz=5, label_title='region') + coord_fixed() + labs(y=exp_names[i]) + theme(text = element_text(size = 20))) %>% get_legend() %>% as_ggplot() 

    p = p + (plot_spatial(c_dt[, .(x, y)], labels=factor(c_dt$grid_2), hide_legend = TRUE, title='2 * 2', sz=2, label_cols=grid2_cols) + coord_fixed() + theme(text = element_text(size = 20))) 

    p = p + (plot_spatial(c_dt[, .(x, y)], labels=factor(c_dt$grid_3), hide_legend = TRUE, title='3 * 3', sz=2, label_cols=grid3_cols) + coord_fixed() + theme(text = element_text(size = 20))) 

    legend_grid = (plot_spatial(c_dt[, .(x, y)], labels=factor(c_dt$grid_3), hide_legend = FALSE, title='3 * 3', sz=5, label_cols=grid3_cols, label_title = 'partition in 3*3 grid') + coord_fixed() + theme(text = element_text(size = 20)) + scale_color_manual(values=grid3_cols, na.value='#e6e6e6', drop=TRUE, limits=names(grid_code), labels=grid_code[names(grid3_cols)])) %>% get_legend() %>% as_ggplot() 

    p = p + (plot_spatial(c_dt[, .(x, y)], labels=factor(c_dt$grid_4), hide_legend = TRUE, title='4 * 4', sz=2, label_cols=grid4_cols) + coord_fixed() + theme(text = element_text(size = 20)))

    p = p + (plot_2d( 
      dim_df = map_2d[['sagenet']][[exp_names[i]]][c_dt$cell_id, ],
      labels = factor(as.character(c_dt$class__)),
      hide_legend = TRUE,
      title = 'SageNet',
      sz = 2, 
      label_cols=class__cols
    ) + theme(text = element_text(size = 20)) )

  }
  else{
    p = p + (plot_spatial(c_dt[, .(x, y)], labels=factor(as.character(c_dt$class__)), hide_legend = TRUE, title='', label_cols=class__cols, sz=2) + coord_fixed() + labs(y=exp_names[i])) + theme(text = element_text(size = 20))

    p = p + (plot_spatial(c_dt[, .(x, y)], labels=factor(c_dt$grid_2), hide_legend = TRUE, title='', sz=2, label_cols=grid2_cols) + coord_fixed())

    p = p + (plot_spatial(c_dt[, .(x, y)], labels=factor(c_dt$grid_3), hide_legend = TRUE, title='', sz=2, label_cols=grid3_cols) + coord_fixed())

    p = p + (plot_spatial(c_dt[, .(x, y)], labels=factor(c_dt$grid_4), hide_legend = TRUE, title='', sz=2, label_cols=grid4_cols) + coord_fixed()) 

    p = p + (plot_2d( 
      dim_df = map_2d[['sagenet']][[exp_names[i]]][c_dt$cell_id, ],
      labels = factor(as.character(c_dt$class__)),
      hide_legend = TRUE,
      title = '',
      sz = 2, 
      label_cols=class__cols
    ) ) 
  }
}

fig_list[['4A']] = p + plot_layout(ncol= 5, guides = 'collect') 
fig_list[['4A']]

legend_region + legend_grid + plot_layout(ncol= 1) 
```

## 4B-D: Benchmark plots; dist. correlation, lisi scores, and cell2cell affinity dist.
```{r 4B_D, fig.height=8, fig.width=31, message=FALSE, cache=FALSE}
fig_list[['4B']] = cor_dt[method != 'sagenet_markers'] %>% 
  ggplot +
  aes(x = sample, y = corr, group = method, 
    color=factor(method, ordered=TRUE, levels=method_ord)) +
  geom_line(size=3) +
  geom_point(size=4) +
  scale_color_manual(values=method_cols, na.value='#e6e6e6', drop=TRUE, 
    limits=setdiff(method_ord, c('True Space')), 
    labels=method_code[!names(method_code) %in% c('True Space')]) +
  labs(color='method', y = "Spearman's rank correlation", x = 'query dataset') +
  theme_bw() +
  theme(text = element_text(size = 30)) + 
  theme(axis.text.x=element_text(angle = 45, hjust = 1)) 


fig_list[['4C']] = ccc_dt[method != 'True Space' & method != 'sagenet_markers'] %>% 
  ggplot +
  aes(x = sample, y = corr, group = method, 
    color=factor(method, ordered=TRUE, levels=method_ord)) +
  # geom_bar(stat="identity", position=position_dodge()) +
  geom_line(size=3) +
  geom_point(size=4) +
  scale_color_manual(values=method_cols, na.value='#e6e6e6', drop=TRUE, 
    limits=setdiff(method_ord, c('True Space')), 
    labels=method_code[!names(method_code) %in% c('True Space')]) +
  labs(color='method', y = "cell-type affinity distance", x = 'query dataset') +
  theme_bw() +
  theme(legend.position='none',
    text = element_text(size = 30),
    axis.text.x=element_text(angle = 45, hjust = 1)) 

# lisi_dists = round(lisi_dists,digits=2)
fig_list[['4D']] = lisi_vals[method != 'sagenet_markers'] %>%
  ggplot +
  aes(x = factor(method, ordered=TRUE, levels=method_ord), y = log2(LISI), 
    fill=factor(method, ordered=TRUE, levels=method_ord)) +
  geom_boxplot(color='#959595') +
  theme_bw() +
  scale_fill_manual(values=method_cols, na.value='#e6e6e6', drop=TRUE, 
    limits=unique(lisi_vals$method)) +
  labs(y='log-scaled LISI (cell type)', fill = 'method', x= 'method') +
  scale_color_manual(values=method_cols, na.value='#e6e6e6') + 
  theme(text = element_text(size = 30), 
    legend.position='none', 
    axis.text.x=element_text(angle = 45, hjust = 1)) + 
  geom_hline(yintercept=0, linetype="dashed", color = "black", size=2) +
  scale_x_discrete(labels=method_code)

fig_list[['4B-D']]  = 
  fig_list[['4B']] + 
  plot_spacer() + 
  fig_list[['4C']] +
  plot_spacer() +  
  fig_list[['4D']] + 
  plot_layout(ncol = 5, widths=c(9, 1, 9, 1, 8), guides='collect')
fig_list[['4B-D']]
```

## Supp12: Benchmark plots; dist. correlation, lisi scores, and cell2cell affinity dist.
```{r Supp12, fig.height=8, fig.width=31, message=FALSE, cache=FALSE}
fig_list[['Supp12A']] = cor_dt %>% 
  ggplot +
  aes(x = sample, y = corr, group = method, 
    color=factor(method, ordered=TRUE, levels=method_ord)) +
  geom_line(size=3) +
  geom_point(size=4) +
  scale_color_manual(values=method_cols, na.value='#e6e6e6', drop=TRUE, 
    limits=setdiff(method_ord, c('True Space')), 
    labels=method_code[!names(method_code) %in% c('True Space')]) +
  labs(color='method', y = "Spearman's rank correlation", x = 'query dataset') +
  theme_bw() +
  theme(text = element_text(size = 30)) + 
  theme(axis.text.x=element_text(angle = 45, hjust = 1)) 


fig_list[['Supp12B']] = ccc_dt[method != 'True Space'] %>% 
  ggplot +
  aes(x = sample, y = corr, group = method, 
    color=factor(method, ordered=TRUE, levels=method_ord)) +
  # geom_bar(stat="identity", position=position_dodge()) +
  geom_line(size=3) +
  geom_point(size=4) +
  scale_color_manual(values=method_cols, na.value='#e6e6e6', drop=TRUE, 
    limits=setdiff(method_ord, c('True Space')), 
    labels=method_code[!names(method_code) %in% c('True Space')]) +
  labs(color='method', y = "cell-type affinity distance", x = 'query dataset') +
  theme_bw() +
  theme(legend.position='none',
    text = element_text(size = 30),
    axis.text.x=element_text(angle = 45, hjust = 1)) 

# lisi_dists = round(lisi_dists,digits=2)
fig_list[['Supp12C']] = lisi_vals %>%
  ggplot +
  aes(x = factor(method, ordered=TRUE, levels=method_ord), y = log2(LISI), 
    fill=factor(method, ordered=TRUE, levels=method_ord)) +
  geom_boxplot(color='#959595') +
  theme_bw() +
  scale_fill_manual(values=method_cols, na.value='#e6e6e6', drop=TRUE, 
    limits=unique(lisi_vals$method)) +
  labs(y='log-scaled LISI (cell type)', fill = 'method', x= 'method') +
  scale_color_manual(values=method_cols, na.value='#e6e6e6') + 
  theme(text = element_text(size = 30), 
    legend.position='none', 
    axis.text.x=element_text(angle = 45, hjust = 1)) + 
  geom_hline(yintercept=0, linetype="dashed", color = "black", size=2) +
  scale_x_discrete(labels=method_code)

fig_list[['Supp12']]  = 
  fig_list[['Supp12A']] + 
  plot_spacer() + 
  fig_list[['Supp12B']] +
  plot_spacer() +  
  fig_list[['Supp12C']] + 
  plot_layout(ncol = 5, widths=c(9, 1, 9, 1, 8), guides='collect')
fig_list[['Supp12']]
```


# Save objects & plots
```{r save, fig.height=12, fig.width=12, message=FALSE, cache=FALSE}
names(obj_list) %>% map(~saveRDS(obj_list[[.x]], file.path(obj_dir, paste0(.x, '.Rds'))))
names(fig_list) %>% map(~saveRDS(fig_list[[.x]], file.path(fig_dir, paste0(.x, '.Rds'))))
```
