---
title: "Figure 2: Spatial Mouse Atlas - Seqfish - Mouse Embryo"
author:
- name: Elyas Heidari
  affiliation: 
  - &IMLS Institute for Molecular Life Sciences, University of Zurich, Switzerland
  - Swiss Institute of Bioinformatics (SIB), University of Zurich, Switzerland
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output:
  workflowr::wflow_html:
    code_folding: hide
    toc: true
    toc_float: true
    number_sections: false
---

# Setup / definitions


```{r setup_knitr, include=FALSE}
library('BiocStyle')
set.seed(1996)
options(bitmapType='cairo')
knitr::opts_chunk$set(autodep=TRUE, cache=FALSE, dev='pdf', cache.lazy = FALSE)
# knitr::opts_chunk$set( autodep=TRUE, cache=TRUE, cache.lazy=FALSE, dev='png' )
knitr::opts_knit$set( root.dir='..' )
# wflow_build(files='analysis/MGA_data_multi_simulate.rmd', view=F, verbose=T, delete_cache=T)
```

```{r setup_libs, collapse=FALSE, message=FALSE, warning=FALSE, cache=FALSE}
```

## Libraries

```{r libs, message=FALSE, warning=FALSE, cache=FALSE}
# library('pika')
library('tidyverse')
library('igraph')
library('data.table')
library('SingleCellExperiment')
library('magrittr')
library('BiocParallel')
library('patchwork')
library('limma')
library('cccd')
library('philentropy')
library('cowplot')
library('zellkonverter')
library('stringr')
library('lisi')
library('ggpubr')
library('Rtsne')
# library('anndata')
```


## Helper functions
```{r setup_helpers, message=FALSE, cache=FALSE}
source('code/utils/utils.R')
source('code/utils/SMA_functions.R')
```

## Directories
```{r setup_input, message=FALSE, cache=FALSE}
tag       = 'seqfish_mouse_embryo/sagenet' 
ref_tags   = c('embryo1_2', 'embryo2_2', 'embryo3_2')
out_dir   = file.path('output', tag)
obj_dir   = 'int_data/seqfish_mouse_embryo/multi'
dir.create(obj_dir)
obj_list  = list()
fig_dir   = 'figures/seqfish_mouse_embryo/multi'
dir.create(fig_dir)
fig_list = list()

names(ref_tags) = ref_tags
preds_f = ref_tags %>% map(~file.path(out_dir, sprintf('preds_%s_query_data.h5ad', .x)))
col_dt_f = file.path('int_data/seqfish_mouse_embryo/', 'embryo1_2', 'col_dt_embryo1_2.txt')
```





## Params
```{r setup_outputs, message=FALSE, cache=FALSE}
overwrite = FALSE
```

# Load inputs
```{r load_inputs, message=FALSE, cache=FALSE}
sce        = preds_f[['embryo1_2']] %>% readH5AD 
cell_idx = which(sce$embryo != 'scRNAseq')
sce_query  = sce[, cell_idx]
meta_query = colData(sce_query) %>% as.data.table
rm(sce)
if(overwrite | 
  !file.exists(file.path(obj_dir, 'map_2d.Rds')) | 
  !file.exists(file.path(obj_dir, 'meta_dt.Rds'))
){
  # Gene importanc
  preds_list = ref_tags %>% purrr::map(~(anndata::read_h5ad(preds_f[[.x]])$obsm$dist_map) %>% as('sparseMatrix') %>% .[cell_idx, cell_idx]) 
gc()
}
```

# Processing / calculations
## Distance matrices and embeddings
```{r calc_dists, message=FALSE, cache=FALSE}
if(overwrite | 
  !file.exists(file.path(obj_dir, 'map_2d.Rds')) | 
  !file.exists(file.path(obj_dir, 'meta_dt.Rds'))
){
  d      = list()
  d_mtx  = list()
  map_2d = list()

  map_2d[['true']]          = as.matrix(meta_query[, .(x,y)])
  rownames(map_2d[['true']]) = meta_query$cell_id

  # sagenet distances
  for(i in ref_tags){
    d[[i]] = preds_list[[i]] %>% as.dist
    d_mtx[[i]] = d[[i]] %>% as.matrix 
    d_mtx[[i]] = d_mtx[[i]]/norm(d_mtx[[i]])
    rownames(d_mtx[[i]]) = colnames(d_mtx[[i]]) = colnames(sce_query)
    map_2d[[i]] = Rtsne(d_mtx[[i]], is_distance=T, perplexity=20)$Y
    # map_2d[[i]] = uwot::umap(d[[i]])
    rownames(map_2d[[i]]) = colnames(sce_query)
    meta_query[[paste0(i, '_1')]] = map_2d[[i]][meta_query$cell_id, 1]
    meta_query[[paste0(i, '_2')]] = map_2d[[i]][meta_query$cell_id, 2]
    # meta_query[, `:=`(i = ..map_2d[[i]][cell_id, 1], 
    #               i = ..map_2d[[i]][cell_id, 2])]   
    preds_list[[i]] = NULL
  }

  d_mtx[['all']] = (d_mtx %>% purrr::reduce(`+`))/3
  rownames(d_mtx[['all']]) = colnames(d_mtx[['all']]) = colnames(sce_query)
  map_2d[['all']] = Rtsne(d_mtx[['all']], is_distance=T, perplexity=20)$Y
  # map_2d[['all']] = uwot::umap(as.dist(d[[i]]))
  rownames(map_2d[['all']]) = colnames(sce_query)
  meta_query[, `:=`(sagenet_all_1 = ..map_2d[['all']][cell_id, 1], 
              sagenet_all_2 = ..map_2d[['all']][cell_id, 2])]   

  # true physical distances
  d_true = dist(meta_query[, .(x,y)])
  d_mtx_true = d_true %>% as.matrix
  rownames(d_mtx_true) = colnames(d_mtx_true) = meta_query$cell_id

  # obj_list[['d_mtx']] = append(d_mtx, d_mtx_true)
  obj_list[['map_2d']] = map_2d
  obj_list[['meta_dt']] = meta_query
  gc()
}else{
  map_2d = readRDS(file.path(obj_dir, 'map_2d.Rds'))
  meta_query = readRDS(file.path(obj_dir, 'meta_dt.Rds'))
}

methods = c('true', 'all', 'embryo3_2', 'embryo2_2', 'embryo1_2')
method_cols = c('#E1C239', "#1965B0", '#FF7F00', "#FDB462", "#FB8072")
method_ord = names(method_cols) = names(methods) = methods
```

## Correlation between the true distance matrix and the prediction distance matrix
```{r calc_cor, message=FALSE, cache=FALSE}
if(overwrite | 
  !file.exists(file.path(obj_dir, 'cor_list.Rds'))
){
  cor_list    = list()
  for(meth in setdiff(names(map_2d), 'true')){

  # for(meth in 'sagenet_markers'){
    cor_list[[meth]]    = c()
    for(e in levels(factor(meta_query$embryo))){
      cells = meta_query[embryo == e]$cell_id
      cor_list[[meth]][e] = cor(c(d_mtx_true[cells, cells]), c(d_mtx[[meth]][cells, cells]), method='spearman')
    }
  }

  obj_list[['cor_list']] = cor_list
  gc()
}else{
  cor_list = readRDS(file.path(obj_dir, 'cor_list.Rds'))
}

cor_dt = names(cor_list) %>%
  map(~data.table(method=.x, embryo=names(cor_list[[.x]]), corr=cor_list[[.x]])) %>%
  purrr::reduce(rbind)
```

## Cell2Cell afinity scores
```{r calc_c2c, message=FALSE, cache=FALSE}
if(overwrite | 
  !file.exists(file.path(obj_dir, 'ccc_list.Rds'))
){
  ccc_list    = list()
  # True physical space
  for(e in levels(factor(meta_query$embryo))){
      cells = meta_query[embryo == e]$cell_id
      g_out    = map_2d[['true']] %>% 
        .[cells, ] %>% 
        get_delaunay(plot=FALSE) %>%
        .$graph
      ccc_list[['true']][[e]] = cellCellContact(
          sce = sce_query[, cells],
          group = 'cell_type',
          graph = g_out,
          nperm = 500,
          plot = FALSE,
          cellID='cell_id')
  }

  # Predicted spaces

  for(meth in setdiff(names(map_2d), 'true')){
  # for(meth in 'sagenet_markers'){
    ccc_list[[meth]] = list()
      for(e in levels(factor(meta_query$embryo))){
      cells = meta_query[embryo == e]$cell_id
      g_out    = map_2d[[meth]] %>% 
        .[cells, ] %>% 
        get_delaunay(plot=FALSE) %>%
        .$graph
      ccc_list[[meth]][[e]] = cellCellContact(
          sce    = sce_query[, cells],
          group  = 'cell_type',
          graph  = g_out,
          nperm  = 500,
          plot   = FALSE,
          cellID ='cell_id')
    }
  }

  obj_list[['ccc_list']] = ccc_list
}else{
  ccc_list = readRDS(file.path(obj_dir, 'ccc_list.Rds'))
}
ccc_dist    = list()
for(meth in setdiff(names(map_2d), 'true')){
  ccc_dist[[meth]] = c()
  for(e in levels(factor(meta_query$embryo))){
    m1 = ccc_list[[meth]][[e]]$pmat 
    m2 = ccc_list[['true']][[e]]$pmat
    ccc_dist[[meth]][e] = norm(m1 - m2 , '2')
  }
}

# Scores' data table 
ccc_dt = names(ccc_dist) %>%
  map(~names(ccc_dist[[.x]]) %>% map(function(y) 
      data.table(method=.x, embryo=y, corr=ccc_dist[[.x]][y])
    ) %>% purrr::reduce(rbind) 
  ) %>%
  purrr::reduce(rbind) 


```


## Distribution of the lisi scores
```{r calc_lisi, message=FALSE, cache=FALSE}
preds_list = list()
if(overwrite | 
  !file.exists(file.path(obj_dir, 'lisi_vals.Rds'))
){
  preds_list = map_2d
  preds_list[['true']] = meta_query[, .(x,y)]
  # lisi_list   = list()
  lisi_vals = names(preds_list) %>%
    map(~compute_lisi(preds_list[[.x]], meta_query, c("class_", "cell_type")) %>%
      as.data.table %>% 
      .[, `:=`(cell_id = ..meta_query$cell_id, method = .x)]) %>%
      purrr::reduce(rbind)
  obj_list[['lisi_vals']] = lisi_vals
}else{
  lisi_vals = readRDS(file.path(obj_dir, 'lisi_vals.Rds'))
}
methods = setdiff(unique(lisi_vals$method), 'true')
names(methods) = methods
lisi_vals %<>% setkey(cell_id)
lisi_vals = lisi_vals[order(cell_id)] %>%
  setnames('cell_type', 'LISI') %>%
  .[meta_query[, .(cell_id, cell_type, embryo)]]
lisi_dists = methods %>%
  map(~dist(rbind(lisi_vals[method == .x]$LISI, lisi_vals[method == 'true']$LISI)), na.rm=T) %>%
  map(c) %>%
  unlist 
```

## A: Benchmark plots; dist. correlation, lisi scores, and cell2cell affinity dist.
```{r A, fig.height=4, fig.width=9, message=FALSE, cache=FALSE}
colblu <- colorRampPalette(c("white", "#00bfff")) 

fig_list[['C1']] = cor_dt %>%
  ggplot +
  aes(x = embryo, fill=corr, y=factor(method, ordered=TRUE, levels=method_ord)) +
  scale_fill_gradientn(colors = colblu(50), breaks = c(min(cor_dt$corr), max(cor_dt$corr)), labels = c("low", "high"))+
  geom_tile() +
  geom_text(aes(label=round(corr,digits=2))) +
  theme_minimal() +
  theme(plot.title = element_text(hjust = 0.5), legend.position='bottom', axis.text.x=element_text(angle = 45, hjust = 1)) +
  labs(title="Spearman's correlation", x='query dataset', y= 'reference dataset', fill="") 

colred <- colorRampPalette(c("#A10037", "white")) 
fig_list[['C2']] = ccc_dt %>%
  ggplot +
  aes(x = embryo, fill=corr, y=factor(method, ordered=TRUE, levels=method_ord)) +
  scale_fill_gradientn(colors = colred(50), breaks = c(min(ccc_dt$corr), max(ccc_dt$corr)), labels = c("low", "high"))+
  geom_tile() +
  geom_text(aes(label=round(corr,digits=2))) +
  theme_minimal() +
  theme(plot.title = element_text(hjust = 0.5), legend.position='bottom', axis.text.x=element_text(angle = 45, hjust = 1), axis.ticks.y=element_blank(), axis.text.y=element_blank()) +
  labs(title="cell-cell affinity distance", x='query dataset', y= '', fill="") 


fig_list[['C']]  = fig_list[['C1']] + fig_list[['C2']] + 
  plot_layout(ncol = 2)
fig_list[['C']] 

# fig_list[['C1']] = cor_dt %>% 
#   ggplot +
#   aes(x = embryo, y = corr, group = method, color=factor(method, ordered=TRUE, levels=method_ord)) +
#   # geom_bar(stat="identity", position=position_dodge()) +
#   geom_line(size=3) +
#   geom_point(size=4) +
#   scale_color_manual(values=method_cols, na.value='gray', drop=TRUE, levels=unique(cor_dt$method)) +
#   labs(color='reference', y = 'spearman correlation') +
#   theme_bw()  + theme(text = element_text(size = 20)) + 
#   theme(axis.text.x=element_text(angle = 45, hjust = 1)) 

# fig_list[['C2']] = ccc_dt %>% 
#   ggplot +
#   aes(x = embryo, y = corr, group = method, color=factor(method, ordered=TRUE, levels=method_ord)) +
#   # geom_bar(stat="identity", position=position_dodge()) +
#   geom_line(size=3) +
#   geom_point(size=4) +
#   theme_bw() +
#   scale_color_manual(values=method_cols, na.value='gray', drop=TRUE) +
#   labs(y = 'cell2cell affinity distance', fill = 'reference') +
#   theme(legend.position='none')  + theme(text = element_text(size = 20)) + 
#   theme(axis.text.x=element_text(angle = 45, hjust = 1))   

# lisi_dists = round(lisi_dists,digits=2)
# fig_list[['C3']] = lisi_vals %>%
#   ggplot +
#   aes(x = factor(method, ordered=TRUE, levels=method_ord), y = LISI, fill=factor(method, ordered=TRUE, levels=method_ord)) +
#   geom_boxplot() +
#   theme_bw() +
#   scale_fill_manual(values=method_cols, na.value='gray') +
#   labs(y='LISI (cell type)', fill = 'reference', x= 'reference dataset') +
#   theme(legend.position='none') +
#   geom_text(aes(x=method, y=max(LISI), label=lisi_dists[method], col=method), size=5) +
#   scale_color_manual(values=method_cols, na.value='gray') + theme(text = element_text(size = 20)) + 
#   theme(axis.text.x=element_text(angle = 45, hjust = 1)) 

# # fig_list[['C']]  = fig_list[['C1']] + fig_list[['C2']] + fig_list[['C3']] + plot_layout(ncol = 3, guides = "collect")

# fig_list[['C']]  = fig_list[['C1']] + fig_list[['C2']] + 
#   plot_layout(ncol = 2, guides = "collect")
# fig_list[['C']]
```

## XA: 
```{r XA, fig.height=10, fig.width=20, message=FALSE, cache=FALSE}
# col_dt  = col_dt_f %>%  fread 
# tab_dt = col_dt[, .N, by = 'cell_type'] %>% .[order(N), ]
# celltype_ord = tab_dt$cell_type
# # lisi_dists = round(lisi_dists,digits=2)
# # lisi_vals %<>% .[method %in% c('true', 'sagenet')]
# fig_list[['XA']] = lisi_vals %>%
#   ggplot +
#   aes(x = factor(cell_type, ordered=TRUE, levels=celltype_ord), y = LISI, color=factor(cell_type, ordered=TRUE, levels=celltype_ord), fill=factor(method, ordered=TRUE, levels=method_ord)) +
#   geom_boxplot() +
#   theme_bw() +
#   scale_color_manual(values=celltype_colours, na.value='gray') +
#   scale_fill_manual(values=method_cols, na.value='gray') +
#   labs(y='LISI (cell type)', fill = 'cell type', x= '') +
#   theme(legend.position='none') + 
#   theme(axis.text.x=element_text(angle = 45, hjust = 1))  + theme(text = element_text(size = 20))   

# fig_list[['XA']]

# fig_list[['XB']] = lisi_vals %>%
#   ggplot +
#   aes(x = embryo, y = LISI, color=embryo, fill=factor(method, ordered=TRUE, levels=method_ord)) +
#   geom_boxplot() +
#   theme_bw() +
#   scale_color_manual(values=.palette1, na.value='gray') +
#   scale_fill_manual(values=method_cols, na.value='gray') +
#   labs(y='LISI (cell type)', fill = 'cell type', x= '') +
#   theme(legend.position='none') + 
#   theme(axis.text.x=element_text(angle = 45, hjust = 1))  + theme(text = element_text(size = 20))   

# fig_list[['XB']]
```




# Save objects & plots
```{r save, fig.height=12, fig.width=12, message=FALSE, cache=FALSE}
names(obj_list) %>% map(~saveRDS(obj_list[[.x]], file.path(obj_dir, paste0(.x, '.Rds'))))
names(fig_list) %>% map(~saveRDS(fig_list[[.x]], file.path(fig_dir, paste0(.x, '.Rds'))))
```
