---
title: "Figure 2: Spatial Mouse Atlas - Seqfish - Mouse Embryo"
author:
- name: Elyas Heidari
  affiliation: 
  - &IMLS Institute for Molecular Life Sciences, University of Zurich, Switzerland
  - Swiss Institute of Bioinformatics (SIB), University of Zurich, Switzerland
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output:
  workflowr::wflow_html:
    code_folding: hide
    toc: true
    toc_float: true
    number_sections: false
---

# Setup / definitions


```{r setup_knitr, include=FALSE}
library('BiocStyle')
set.seed(1996)
options(bitmapType='cairo')
knitr::opts_chunk$set(autodep=TRUE, cache=FALSE, dev='pdf', cache.lazy = FALSE)
# knitr::opts_chunk$set( autodep=TRUE, cache=TRUE, cache.lazy=FALSE, dev='png' )
knitr::opts_knit$set( root.dir='..' )
# wflow_build(files='analysis/MGA_data_multi_all.rmd', view=F, verbose=T, delete_cache=T)
```

```{r setup_libs, collapse=FALSE, message=FALSE, warning=FALSE, cache=FALSE}
```

## Libraries

```{r libs, message=FALSE, warning=FALSE, cache=FALSE}
# library('pika')
library('tidyverse')
library('igraph')
library('data.table')
library('SingleCellExperiment')
library('magrittr')
library('BiocParallel')
library('patchwork')
library('limma')
library('cccd')
library('philentropy')
library('cowplot')
library('zellkonverter')
library('stringr')
library('lisi')
library('ggpubr')
library('Rtsne')
# library('anndata')
```


## Helper functions
```{r setup_helpers, message=FALSE, cache=FALSE}
source('code/utils/utils.R')
source('code/utils/SMA_functions.R')
```

## Directories
```{r setup_input, message=FALSE, cache=FALSE}
tag       = 'seqfish_mouse_embryo/sagenet' 
ref_tags   = c('embryo1_2', 'embryo2_2', 'embryo3_2')
query_tag  = c('scRNAseq', 'embryo1_5', 'embryo2_5', 'embryo3_5')
out_dir   = file.path('output', tag)
obj_dir   = 'int_data/seqfish_mouse_embryo/all'
dir.create(obj_dir)
obj_list  = list()
fig_dir   = 'figures/seqfish_mouse_embryo/all'
dir.create(fig_dir)
fig_list = list()

names(ref_tags) = ref_tags
preds_f = ref_tags %>% map(~file.path(out_dir, sprintf('preds_%s_query_data.h5ad', .x)))
col_dt_f = file.path('int_data/seqfish_mouse_embryo/', 'embryo1_2', 'col_dt_embryo1_2.txt')
```





## Params
```{r setup_outputs, message=FALSE, cache=FALSE}
overwrite = FALSE
```

# Load inputs
```{r load_inputs, message=FALSE, cache=FALSE}
sce_query = preds_f[['embryo1_2']] %>% readH5AD 
cell_idx  = which(sce_query$embryo %in% query_tag)
sce_query = sce_query[, cell_idx]
meta_query = colData(sce_query) %>% as.data.table
# rm(sce_query)
ent = preds_f %>% map(readH5AD) %>% map(get_confidence) %>% purrr::reduce(`+`) %>% .[cell_idx]
meta_query$ent = ent / max(ent)
gc()
if(overwrite | 
  !file.exists(file.path(obj_dir, 'map_2d.Rds')) | 
  !file.exists(file.path(obj_dir, 'meta_dt.Rds'))
){
  # Gene importanc
  preds_list = ref_tags %>% purrr::map(~(anndata::read_h5ad(preds_f[[.x]])$obsm$dist_map) %>% as('sparseMatrix') %>% .[cell_idx, cell_idx]) 
gc()
}
```

# Processing / calculations
## Distance matrices and embeddings
```{r calc_dists, message=FALSE, cache=FALSE}
if(overwrite | 
  !file.exists(file.path(obj_dir, 'map_2d.Rds')) | 
  !file.exists(file.path(obj_dir, 'meta_dt.Rds'))
){
  d      = list()
  d_mtx  = list()
  map_2d = list()

  map_2d[['true']]          = as.matrix(meta_query[, .(x,y)])
  rownames(map_2d[['true']]) = meta_query$cell_id

  # sagenet distances
  for(i in ref_tags){
    d[[i]] = preds_list[[i]] %>% as.dist
    d_mtx[[i]] = d[[i]] %>% as.matrix 
    d_mtx[[i]] = d_mtx[[i]]/norm(d_mtx[[i]])
    rownames(d_mtx[[i]]) = colnames(d_mtx[[i]]) = colnames(sce_query)
    map_2d[[i]] = Rtsne(d_mtx[[i]], is_distance=T, perplexity=50)$Y
    # map_2d[[i]] = uwot::umap(d[[i]])
    rownames(map_2d[[i]]) = colnames(sce_query)
    meta_query[[paste0(i, '_1')]] = map_2d[[i]][meta_query$cell_id, 1]
    meta_query[[paste0(i, '_2')]] = map_2d[[i]][meta_query$cell_id, 2]
    # meta_query[, `:=`(i = ..map_2d[[i]][cell_id, 1], 
    #               i = ..map_2d[[i]][cell_id, 2])]   
    preds_list[[i]] = NULL
  }

  d_mtx[['all']] = (d_mtx %>% purrr::reduce(`+`))/3
  rownames(d_mtx[['all']]) = colnames(d_mtx[['all']]) = colnames(sce_query)
  map_2d[['all']] = Rtsne(d_mtx[['all']], is_distance=T, perplexity=50)$Y
  # map_2d[['all']] = uwot::umap(as.dist(d[[i]]))
  rownames(map_2d[['all']]) = colnames(sce_query)
  meta_query[, `:=`(sagenet_all_1 = ..map_2d[['all']][cell_id, 1], 
              sagenet_all_2 = ..map_2d[['all']][cell_id, 2])] 

  d[['exprs']] = dist(t(assays(sce_query)[[1]])) %>% as.dist
  d_mtx[['exprs']] = d[['exprs']] %>% as.matrix 
  rownames(d_mtx[['exprs']]) = colnames(d_mtx[['exprs']]) = colnames(sce_query)
  map_2d[['exprs']] = Rtsne(d_mtx[['exprs']], is_distance=T, perplexity=50)$Y
  rownames(map_2d[['exprs']]) = colnames(sce_query)
  meta_query[, `:=`(expr_tsne_1 = ..map_2d[['exprs']][cell_id, 1], 
                expr_tsne_2 = ..map_2d[['exprs']][cell_id, 2])] 

  d_true = dist(meta_query[, .(x,y)])
  d_mtx_true = d_true %>% as.matrix
  rownames(d_mtx_true) = colnames(d_mtx_true) = meta_query$cell_id
  # obj_list[['d_mtx']] = append(d_mtx, d_mtx_true)
  obj_list[['map_2d']] = map_2d
  obj_list[['meta_dt']] = meta_query
  gc()
}else{
  map_2d = readRDS(file.path(obj_dir, 'map_2d.Rds'))
  meta_query = readRDS(file.path(obj_dir, 'meta_dt.Rds'))
}

methods = c('true', 'all', 'embryo1_2', 'embryo2_2', 'embryo3_2')
method_cols = c('#E1C239', "#1965B0", '#FF7F00', "#FDB462", "#FB8072")
method_ord = names(method_cols) = names(methods) = methods
```


## Cell2Cell afinity scores
```{r calc_c2c, fig.height=15, fig.width=15, message=FALSE, cache=FALSE}

g_out    = map_2d[['all']] %>%
  get_delaunay(plot=FALSE) %>%
  .$graph

ccc_map = cellCellContact(
          sce    = sce_query,
          group  = 'cell_type',
          graph  = g_out,
          nperm  = 500,
          plot   = FALSE,
          cellID ='cell_id')

obj_list[['ccc_map']] = map_2d

ccc_map = cellCellContactHeatmapTriangle(ccc_map, celltype_colours, title='')

ccc_map
```


## B: UMAP plosts of expression space and the reconstructed space
```{r B, fig.height=10, fig.width=11, message=FALSE, cache=FALSE, eval=TRUE}
meta_query %<>% .[, .SD[sample(.N)]]
meta_query$ent = 1 - meta_query$ent/(max(meta_query$ent))
exprs_p = plot_2d( 
    dim_df = meta_query[, .(expr_tsne_1, expr_tsne_2)],
    labels = meta_query$cell_type,
    label_cols = celltype_colours,
    hide_legend = TRUE,
    title = 'Expression',
    sz = 1
) + labs(x = '', y = '')



sagenet_p = plot_2d( 
    dim_df = meta_query[, .(sagenet_all_1, sagenet_all_2)],
    labels = meta_query$cell_type,
    label_cols = celltype_colours,
    hide_legend = TRUE,
    title = 'SageNet',
    sz = 1,
    alpha = meta_query$ent,
) + labs(x = '', y = '')

exprs_batch_p = plot_2d( 
    dim_df = meta_query[, .(expr_tsne_1, expr_tsne_2)],
    labels = meta_query$embryo,
    label_cols = .palette1,
    hide_legend = TRUE,
    title = '',
    sz = 1
) + labs(x = '', y = '')

sagenet_batch_p = plot_2d( 
    dim_df = meta_query[, .(sagenet_all_1, sagenet_all_2)],
    labels = meta_query$embryo,
    label_cols = .palette1,
    hide_legend = FALSE,
    label_title = 'dataset',
    sz = 1,
    alpha = meta_query$ent,
) + labs(x = '', y = '', alpha='confidence')


fig_list[['B']] = exprs_p + sagenet_p + exprs_batch_p + sagenet_batch_p  + 
  plot_layout(ncol = 2, guides = 'collect')
fig_list[['B']]

```



# Save objects & plots
```{r save, fig.height=12, fig.width=12, message=FALSE, cache=FALSE}
names(obj_list) %>% map(~saveRDS(obj_list[[.x]], file.path(obj_dir, paste0(.x, '.Rds'))))
names(fig_list) %>% map(~saveRDS(fig_list[[.x]], file.path(fig_dir, paste0(.x, '.Rds'))))
```
