---
title: "Figure 2: Spatial Mouse Atlas - Seqfish - Mouse Embryo"
author:
- name: Elyas Heidari
  affiliation: 
  - &IMLS Institute for Molecular Life Sciences, University of Zurich, Switzerland
  - Swiss Institute of Bioinformatics (SIB), University of Zurich, Switzerland
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output:
  workflowr::wflow_html:
    code_folding: hide
    toc: true
    toc_float: true
    number_sections: false
---

# Setup / definitions


```{r setup_knitr, include=FALSE}
library('BiocStyle')
set.seed(1996)
options(bitmapType='cairo')
knitr::opts_chunk$set(autodep=TRUE, cache=FALSE, dev='pdf', cache.lazy = FALSE)
# knitr::opts_chunk$set( autodep=TRUE, cache=TRUE, cache.lazy=FALSE, dev='png' )
knitr::opts_knit$set( root.dir='..' )
# wflow_build(files='analysis/MGA_data_embryo1_2.rmd', view=F, verbose=T, delete_cache=T)
```

```{r setup_libs, collapse=FALSE, message=FALSE, warning=FALSE, cache=FALSE}
```

## Libraries

```{r libs, message=FALSE, warning=FALSE, cache=FALSE}
# library('pika')
library('tidyverse')
library('igraph')
library('data.table')
library('SingleCellExperiment')
library('magrittr')
library('BiocParallel')
library('patchwork')
library('limma')
library('cccd')
library('philentropy')
library('cowplot')
library('zellkonverter')
library('stringr')
library('lisi')
library('ggpubr')
library('Rtsne')
# library('anndata')
```


## Helper functions
```{r setup_helpers, message=FALSE, cache=FALSE}
source('code/utils/utils.R')
source('code/utils/SMA_functions.R')
```

## Directories
```{r setup_input, message=FALSE, cache=FALSE}
tag       = 'seqfish_mouse_embryo' 
ref_tag   = 'embryo1_2'
out_dir   = file.path('output', tag)
obj_dir   = 'int_data/seqfish_mouse_embryo/embryo1_2'
dir.create(obj_dir)
obj_list  = list()
fig_dir   = 'figures/seqfish_mouse_embryo/embryo1_2'
dir.create(fig_dir)
fig_list = list()

methods     = list.files(out_dir)
names(methods) = methods
preds_f   = methods %>%
  map(~file.path(out_dir, .x, paste0(paste('preds', ref_tag, 'query_data', sep='_'), '.h5ad')))
# imp_f = file.path(obj_dir, 'gene_dt_embryo1_2.txt')
imp_f = file.path(obj_dir, 'imp_embryo1_2.txt')
col_dt_f = file.path(obj_dir, 'col_dt_embryo1_2.txt')
```





## Params
```{r setup_outputs, message=FALSE, cache=FALSE}
n_markers = 10
overwrite = FALSE
```

# Load inputs
```{r load_inputs, message=FALSE, cache=FALSE}
sce = preds_f[['sagenet']] %>% readH5AD 
cell_idx = which(sce$embryo != 'scRNAseq')
sce %<>% .[, cell_idx]
meta_dt = colData(sce) %>% as.data.table
if(overwrite | 
  !file.exists(file.path(obj_dir, 'map_2d.Rds')) | 
  !file.exists(file.path(obj_dir, 'meta_dt.Rds'))
){
  # Gene importanc

  preds_list = list()
  preds_list[['sagenet']] = anndata::read_h5ad(preds_f[['sagenet']])$obsm$dist_map 
  adj = anndata::read_h5ad(preds_f[['sagenet']])$varm$adj
  preds_list[['sagenet']]  %<>% as('sparseMatrix') %>% .[cell_idx, cell_idx]
  ## Novosparc predictions
  preds_list[['novosparc']] = preds_f[['novosparc']] %>% read_preds_novosparc %>% .[cell_idx, ]
  gc()
  # Tangram predictions
  preds_list[['tangram_cells']] = preds_f[['tangram_cells']] %>% read_preds_tangram %>% .[cell_idx, ]
  gc()


gc()
}

imp = imp_f %>% fread %>% setnames('V1', 'GENE')
genes = imp$GENE
imp %<>% .[,-1] %>% as.matrix
rownames(imp) = genes
imp_abs = abs(imp)
# imp = apply(imp, 2, function(x) (x)/(max(x)))
imp_dt = apply(imp, 2, function(x) rownames(imp)[order(-x)[1:n_markers]]) %>%
    as.data.table
markers = imp_dt %>% unlist %>% setdiff(c())
print(markers)
col_dt  = col_dt_f %>%  fread 
```

# Processing / calculations
## Distance matrices and embeddings
```{r calc_dists, message=FALSE, cache=FALSE}
if(overwrite | 
  !file.exists(file.path(obj_dir, 'map_2d.Rds')) | 
  !file.exists(file.path(obj_dir, 'meta_dt.Rds'))
){
  d      = list()
  d_mtx  = list()
  map_2d = list()

  map_2d[['tangram_cells']] = preds_list[['tangram_cells']]
  map_2d[['true']]          = as.matrix(meta_dt[, .(x,y)])
  rownames(map_2d[['true']]) = meta_dt$cell_id
  map_2d[['novosparc']]     = preds_list[['novosparc']]

  # sagenet distances
  d[['sagenet']] = preds_list[['sagenet']] %>% as.dist
  d_mtx[['sagenet']] = d[['sagenet']] %>% as.matrix 
  rownames(d_mtx[['sagenet']]) = colnames(d_mtx[['sagenet']]) = colnames(sce)
  map_2d[['sagenet']] = Rtsne(d_mtx[['sagenet']], is_distance=T, perplexity=50)$Y
  rownames(map_2d[['sagenet']]) = colnames(sce)
  meta_dt[, `:=`(sagenet_1 = ..map_2d[['sagenet']][cell_id, 1], 
                sagenet_2 = ..map_2d[['sagenet']][cell_id, 2])]

  # distances based on expression values
  d[['exprs']] = dist(t(assays(sce)[[1]])) %>% as.dist
  d_mtx[['exprs']] = d[['exprs']] %>% as.matrix 
  rownames(d_mtx[['exprs']]) = colnames(d_mtx[['exprs']]) = colnames(sce)
  # map_2d[['exprs']] = uwot::umap(d[['exprs']], n_neighbors=50)
  map_2d[['exprs']] = Rtsne(d_mtx[['exprs']], is_distance=T, perplexity=50)$Y
  rownames(map_2d[['exprs']]) = colnames(sce)
  meta_dt[, `:=`(expr_tsne_1 = ..map_2d[['exprs']][cell_id, 1], 
                expr_tsne_2 = ..map_2d[['exprs']][cell_id, 2])]

  # distances in the marker-subset space
  d[['sagenet_markers']] = dist(t(assays(sce[markers,])[[1]])) %>% as.dist
  d_mtx[['sagenet_markers']] = d[['sagenet_markers']] %>% as.matrix 
  rownames(d_mtx[['sagenet_markers']]) = colnames(d_mtx[['sagenet_markers']]) = colnames(sce)
  # map_2d[['sagenet_markers']] = uwot::umap(d[['sagenet_markers']], n_neighbors=50)
  map_2d[['sagenet_markers']] = Rtsne(d_mtx[['sagenet_markers']], is_distance=T, perplexity=50)$Y
  rownames(map_2d[['sagenet_markers']]) = colnames(sce)
  meta_dt[, `:=`(markers_tsne_1 = ..map_2d[['sagenet_markers']][cell_id, 1], 
                markers_tsne_2 = ..map_2d[['sagenet_markers']][cell_id, 2])]

  # tangram
  d[['tangram_cells']] = dist(preds_list[['tangram_cells']])
  d_mtx[['tangram_cells']] = d[['tangram_cells']] %>% as.matrix
  rownames(d_mtx[['tangram_cells']]) = colnames(d_mtx[['tangram_cells']]) = meta_dt$cell_id
  rownames(map_2d[['tangram_cells']]) = colnames(sce)
  meta_dt[, `:=`(tangram_1 = ..map_2d[['tangram_cells']][cell_id, 1], 
                tangram_2 = ..map_2d[['tangram_cells']][cell_id, 2])]

  d[['novosparc']] = dist(preds_list[['novosparc']])
  d_mtx[['novosparc']] = d[['novosparc']] %>% as.matrix
  rownames(d_mtx[['novosparc']]) = colnames(d_mtx[['novosparc']]) = meta_dt$cell_id
  rownames(map_2d[['novosparc']]) = colnames(sce)
  meta_dt[, `:=`(novosparc_1 = ..map_2d[['novosparc']][cell_id, 1], 
                novosparc_2 = ..map_2d[['novosparc']][cell_id, 2])]

  # true physical distances
  d_true = dist(meta_dt[, .(x,y)])
  d_mtx_true = d_true %>% as.matrix
  rownames(d_mtx_true) = colnames(d_mtx_true) = meta_dt$cell_id

  # obj_list[['d_mtx']] = append(d_mtx, d_mtx_true)
  obj_list[['map_2d']] = map_2d
  obj_list[['meta_dt']] = meta_dt
  gc()
}else{
  map_2d = readRDS(file.path(obj_dir, 'map_2d.Rds'))
  meta_dt = readRDS(file.path(obj_dir, 'meta_dt.Rds'))
}
```

## Cell2Cell afinity scores
```{r calc_c2c, message=FALSE, cache=FALSE}
if(overwrite | 
  !file.exists(file.path(obj_dir, 'ccc_list.Rds'))
){
  ccc_list    = list()
  # True physical space
  for(e in levels(factor(meta_dt$embryo))){
      cells = meta_dt[embryo == e]$cell_id
      g_out    = map_2d[['true']] %>% 
        .[cells, ] %>% 
        get_delaunay(plot=FALSE) %>%
        .$graph
      ccc_list[['true']][[e]] = cellCellContact(
          sce = sce[, cells],
          group = 'cell_type',
          graph = g_out,
          nperm = 500,
          plot = FALSE,
          cellID='cell_id')
  }

  # Predicted spaces

  for(meth in setdiff(names(map_2d), 'true')){
  # for(meth in 'sagenet_markers'){
    ccc_list[[meth]] = list()
      for(e in levels(factor(meta_dt$embryo))){
      cells = meta_dt[embryo == e]$cell_id
      g_out    = map_2d[[meth]] %>% 
        .[cells, ] %>% 
        get_delaunay(plot=FALSE) %>%
        .$graph
      ccc_list[[meth]][[e]] = cellCellContact(
          sce    = sce[, cells],
          group  = 'cell_type',
          graph  = g_out,
          nperm  = 500,
          plot   = FALSE,
          cellID ='cell_id')
    }
  }

  obj_list[['ccc_list']] = ccc_list
}else{
  ccc_list = readRDS(file.path(obj_dir, 'ccc_list.Rds'))
}
ccc_dist    = list()
for(meth in setdiff(names(map_2d), 'true')){
  ccc_dist[[meth]] = c()
  for(e in levels(factor(meta_dt$embryo))){
    m1 = ccc_list[[meth]][[e]]$pmat 
    m2 = ccc_list[['true']][[e]]$pmat
    ccc_dist[[meth]][e] = norm(m1 - m2 , '2')
  }
}

ccc_plot_list = names(ccc_list) %>%
  map(~ccc_list[[.x]][3] %>% map(function(y) cellCellContactHeatmapTriangle(y, celltype_colours, title=.x)))

# Scores' data table 
ccc_dt = names(ccc_dist) %>%
  map(~names(ccc_dist[[.x]]) %>% map(function(y) 
      data.table(method=.x, embryo=y, corr=ccc_dist[[.x]][y])
    ) %>% purrr::reduce(rbind) 
  ) %>%
  purrr::reduce(rbind) 


```


## Correlation between the true distance matrix and the prediction distance matrix
```{r calc_cor, message=FALSE, cache=FALSE}
if(overwrite | 
  !file.exists(file.path(obj_dir, 'cor_list.Rds'))
){
  cor_list    = list()
  for(meth in setdiff(names(map_2d), 'true')){

  # for(meth in 'sagenet_markers'){
    cor_list[[meth]]    = c()
    for(e in levels(factor(meta_dt$embryo))){
      cells = meta_dt[embryo == e]$cell_id
      cor_list[[meth]][e] = cor(c(d_mtx_true[cells, cells]), c(d_mtx[[meth]][cells, cells]), method='spearman')
    }
  }

  obj_list[['cor_list']] = cor_list
  gc()
}else{
  cor_list = readRDS(file.path(obj_dir, 'cor_list.Rds'))
}

cor_dt = names(cor_list) %>%
  map(~data.table(method=.x, embryo=names(cor_list[[.x]]), corr=cor_list[[.x]])) %>%
  purrr::reduce(rbind)
```

## Distribution of the lisi scores
```{r calc_lisi, message=FALSE, cache=FALSE}
preds_list = list()
if(overwrite | 
  !file.exists(file.path(obj_dir, 'lisi_vals.Rds'))
){
  preds_list[['sagenet']] = map_2d[['sagenet']]
  preds_list[['exprs']] = map_2d[['exprs']]
  preds_list[['sagenet_markers']] = map_2d[['sagenet_markers']]
  preds_list[['true']] = meta_dt[, .(x,y)]
  preds_list[['novosparc']] = map_2d[['novosparc']]
  preds_list[['tangram_cells']] = map_2d[['tangram_cells']]
  # lisi_list   = list()
  lisi_vals = names(preds_list) %>%
    map(~compute_lisi(preds_list[[.x]], meta_dt, c("class_", "cell_type")) %>%
      as.data.table %>% 
      .[, `:=`(cell_id = ..meta_dt$cell_id, method = .x)]) %>%
      purrr::reduce(rbind)
  obj_list[['lisi_vals']] = lisi_vals
}else{
  lisi_vals = readRDS(file.path(obj_dir, 'lisi_vals.Rds'))
}
methods = setdiff(unique(lisi_vals$method), 'true')
names(methods) = methods
lisi_vals %<>% setkey(cell_id)
lisi_vals = lisi_vals[order(cell_id)] %>%
  setnames('cell_type', 'LISI') %>%
  .[meta_dt[, .(cell_id, cell_type)]]
lisi_dists = methods %>%
  map(~dist(rbind(lisi_vals[method == .x]$LISI, lisi_vals[method == 'true']$LISI)), na.rm=T) %>%
  map(c) %>%
  unlist 
```

## A: True regional labels + cell types (ref)
```{r A, fig.height=3, fig.width=10, message=FALSE, cache=FALSE}
p = plot_spatial(col_dt[, .(x, -y)], labels=factor(as.character(col_dt$cell_type), ordered=TRUE), label_cols=celltype_colours, hide_legend = TRUE, title = 'cell types') +coord_fixed()
p1 = plot_spatial(col_dt[, .(x, -y)], labels=factor(col_dt$leiden_0.01), hide_legend = TRUE, title='partitioning 1')+coord_fixed()
p2 = plot_spatial(col_dt[, .(x, -y)], labels=factor(col_dt$leiden_0.05), hide_legend = TRUE, title='partitioning 2')+coord_fixed()
p3 = plot_spatial(col_dt[, .(x, -y)], labels=factor(col_dt$leiden_0.1), hide_legend = TRUE, title='partitioning 3')+coord_fixed()
p4 = plot_spatial(col_dt[, .(x, -y)], labels=factor(col_dt$leiden_0.5), hide_legend = TRUE, title='partitioning 4')+coord_fixed()
p5 = plot_spatial(col_dt[, .(x, -y)], labels=factor(col_dt$leiden_1), hide_legend = TRUE, title='partitioning 5')+coord_fixed()

fig_list[['A']] = p + p1 + p2 + p3 + p4 + p5 + plot_layout(nrow=1) 
fig_list[['A']]
# plist
```

## B: UMAP plosts of expression space and the reconstructed space
```{r B, fig.height=10, fig.width=15, message=FALSE, cache=FALSE, eval=TRUE}

meta_dt[, ent :=  (ent_embryo1_2_leiden_1 + ent_embryo1_2_leiden_0.1 + ent_embryo1_2_leiden_0.01 + ent_embryo1_2_leiden_0.5 + ent_embryo1_2_leiden_0.05)/5 ]
meta_dt$ent = 1 - meta_dt$ent/(max(meta_dt$ent))
exprs_p = plot_2d( 
    dim_df = meta_dt[, .(expr_tsne_1, expr_tsne_2)],
    labels = meta_dt$cell_type,
    label_cols = celltype_colours,
    hide_legend = TRUE,
    title = 'Expression',
    sz = 1
) + labs(x = '', y = '')


# lgened = (plot_2d( 
#     dim_df = meta_dt[, .(expr_tsne_1, expr_tsne_2)],
#     labels = meta_dt$cell_type,
#     label_cols = celltype_colours,
#     hide_legend = FALSE,
#     label_title = 'cell type',
#     title = 'Expression',
#     sz = 1
# ) + labs(x = '', y = '')+ theme(text = element_text(size = 10))   ) %>% get_legend() %>%
#   as_ggplot() 


markers_p = plot_2d( 
    dim_df = meta_dt[, .(markers_tsne_1, markers_tsne_2)],
    labels = meta_dt$cell_type,
    label_cols = celltype_colours,
    hide_legend = TRUE,
    title = 'SageNet Markers',
    sz = 1
) + labs(x = '', y = '') 

sagenet_p = plot_2d( 
    dim_df = meta_dt[, .(sagenet_1, sagenet_2)],
    labels = meta_dt$cell_type,
    label_cols = celltype_colours,
    hide_legend = TRUE,
    title = 'SageNet',
    sz = 1,
    alpha = meta_dt$ent,
) + labs(x = '', y = '')

lgened = (plot_2d( 
    dim_df = meta_dt[, .(sagenet_1, sagenet_2)],
    labels = meta_dt$cell_type,
    label_cols = celltype_colours,
    hide_legend = FALSE,
    title = 'SageNet',
    sz = 1,
    alpha = meta_dt$ent,
) + labs(x = '', y = '')+ theme(text = element_text(size = 12))   ) %>% get_legend() %>%
  as_ggplot() 

novosparc_p = plot_2d( 
    dim_df = meta_dt[, .(novosparc_1, novosparc_2)],
    labels = meta_dt$cell_type,
    label_cols = celltype_colours,
    hide_legend = TRUE,
    title = 'novosparc',
    sz = 1
) + labs(x = '', y = '') 

tangram_p = plot_2d( 
    dim_df = meta_dt[, .(tangram_1, tangram_2)],
    labels = meta_dt$cell_type,
    label_cols = celltype_colours,
    hide_legend = TRUE,
    title = 'Tangram',
    sz = 1
) + labs(x = '', y = '') 

fig_list[['B']] = exprs_p + markers_p + sagenet_p + novosparc_p + tangram_p  + lgened +
  plot_layout(ncol = 3)
fig_list[['B']]

```

## C: Benchmark plots; dist. correlation, lisi scores, and cell2cell affinity dist.
```{r C, fig.height=7, fig.width=25, message=FALSE, cache=FALSE}
fig_list[['C1']] = cor_dt %>% 
  ggplot +
  aes(x = embryo, y = corr, group = method, color=factor(method, ordered=TRUE, levels=method_ord)) +
  # geom_bar(stat="identity", position=position_dodge()) +
  geom_line() +
  geom_point() +
  scale_color_manual(values=method_cols, na.value='gray') +
  labs(fill='method', y = 'spearman correlation') +
  theme_bw() +
  theme(legend.position='none')  + theme(text = element_text(size = 20)) + 
  theme(axis.text.x=element_text(angle = 45, hjust = 1)) 

fig_list[['C2']] = ccc_dt %>% 
  ggplot +
  aes(x = embryo, y = corr, group = method, color=factor(method, ordered=TRUE, levels=method_ord)) +
  # geom_bar(stat="identity", position=position_dodge()) +
  geom_line() +
  geom_point() +
  theme_bw() +
  scale_color_manual(values=method_cols, na.value='gray') +
  labs(y = 'cell2cell affinity distance', fill = 'method') +
  theme(legend.position='none')  + theme(text = element_text(size = 20)) + 
  theme(axis.text.x=element_text(angle = 45, hjust = 1))   

lisi_dists = round(lisi_dists,digits=2)
fig_list[['C3']] = lisi_vals %>%
  ggplot +
  aes(x = factor(method, ordered=TRUE, levels=method_ord), y = LISI, fill=factor(method, ordered=TRUE, levels=method_ord)) +
  geom_boxplot() +
  theme_bw() +
  scale_fill_manual(values=method_cols, na.value='gray') +
  labs(y='LISI (cell type)', fill = 'method', x= 'method') +
  theme(legend.position='none') +
  geom_text(aes(x=method, y=max(LISI), label=lisi_dists[method], col=method), size=5) +
  scale_color_manual(values=method_cols, na.value='gray') + theme(text = element_text(size = 20)) + 
  theme(axis.text.x=element_text(angle = 45, hjust = 1)) 

fig_list[['C']]  = fig_list[['C1']] + fig_list[['C2']] + fig_list[['C3']] + plot_layout(ncol = 3)
fig_list[['C']]
```

## XB: LISI scores by cell type
```{r XB, fig.height=10, fig.width=30, message=FALSE, cache=FALSE}
tab_dt = col_dt[, .N, by = 'cell_type'] %>% .[order(N), ]
celltype_ord = tab_dt$cell_type
# lisi_dists = round(lisi_dists,digits=2)
# lisi_vals %<>% .[method %in% c('true', 'sagenet')]
fig_list[['XB']] = lisi_vals %>%
  ggplot +
  aes(x = factor(cell_type, ordered=TRUE, levels=celltype_ord), y = LISI, color=factor(cell_type, ordered=TRUE, levels=celltype_ord), fill=factor(method, ordered=TRUE, levels=method_ord)) +
  geom_boxplot() +
  theme_bw() +
  scale_color_manual(values=celltype_colours, na.value='gray') +
  scale_fill_manual(values=method_cols, na.value='gray') +
  labs(y='LISI (cell type)', fill = 'cell type', x= '') +
  theme(legend.position='none') + 
  theme(axis.text.x=element_text(angle = 45, hjust = 1))  + theme(text = element_text(size = 20))   

fig_list[['XB']]
```

## XC: Cell2Cell affinity plots
```{r XC, fig.height=12, fig.width=12, message=FALSE, cache=FALSE}
fig_list[['XC']] = ccc_plot_list 
fig_list[['XC']]
```

## XD: 
```{r XD, fig.height=7, fig.width=11, message=FALSE, cache=FALSE}
tab_dt = col_dt[, .N, by = 'cell_type'] %>% .[order(N), ]
celltype_ord = tab_dt$cell_type
p1 = tab_dt %>%
  ggplot +
  aes(x = factor(cell_type, ordered=TRUE, levels= celltype_ord), y = N, fill=cell_type) +
  geom_bar(stat="identity") +
  theme_bw() +
  scale_fill_manual(values=celltype_colours, na.value='gray') +
  labs(y='number of cells', x= '') +
  theme(legend.position='none') + 
  theme(axis.text.x=element_blank())

p2 = meta_dt %>%
  ggplot +
  aes(x = factor(cell_type, ordered=TRUE, levels= celltype_ord), y = ent, fill=cell_type) +
  geom_violin() +
  # + geom_jitter(height = 0, width = 0.1) +
  theme_bw() +
  scale_fill_manual(values=celltype_colours, na.value='gray') +
  labs(y='confidence score', x= '') +
  theme(legend.position='none') + 
  theme(axis.text.x=element_text(angle = 45, hjust = 1))

fig_list[['XD']] = p1 + p2 + plot_layout(ncol = 1)
fig_list[['XD']]
```


# Save objects & plots
```{r save, fig.height=12, fig.width=12, message=FALSE, cache=FALSE}
names(obj_list) %>% map(~saveRDS(obj_list[[.x]], file.path(obj_dir, paste0(.x, '.Rds'))))
names(fig_list) %>% map(~saveRDS(fig_list[[.x]], file.path(fig_dir, paste0(.x, '.Rds'))))
```
