---
title: "Figure S1: Spatial Mouse Atlas - Seqfish - Mouse Embryo"
author:
- name: Elyas Heidari
  affiliation: 
  - &IMLS Institute for Molecular Life Sciences, University of Zurich, Switzerland
  - Swiss Institute of Bioinformatics (SIB), University of Zurich, Switzerland
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output:
  workflowr::wflow_html:
    code_folding: hide
    toc: true
    toc_float: true
    number_sections: false
---

# Setup / definitions


```{r setup_knitr, include=FALSE}
library('BiocStyle')
set.seed(1996)
options(bitmapType='cairo')
knitr::opts_chunk$set(autodep=TRUE, cache=FALSE, dev='pdf', cache.lazy = FALSE)
# knitr::opts_chunk$set( autodep=TRUE, cache=TRUE, cache.lazy=FALSE, dev='png' )
knitr::opts_knit$set( root.dir='..' )
# wflow_build(files='analysis/MGA_data_embryo2_2.rmd', view=F, verbose=T, delete_cache=T)
```

```{r setup_libs, collapse=FALSE, message=FALSE, warning=FALSE, cache=FALSE}
```

## Libraries

```{r libs, message=FALSE, warning=FALSE, cache=FALSE}
# library('pika')
library('tidyverse')
library('igraph')
library('data.table')
library('SingleCellExperiment')
library('magrittr')
library('BiocParallel')
library('patchwork')
library('limma')
library('cccd')
library('philentropy')
library('cowplot')
library('zellkonverter')
library('stringr')
library('lisi')
library('ggpubr')
library('Rtsne')
# library('anndata')
```


## Helper functions
```{r setup_helpers, message=FALSE, cache=FALSE}
source('code/utils/utils.R')
source('code/utils/SMA_functions.R')
```

## Directories
```{r setup_input, message=FALSE, cache=FALSE}
tag       = 'seqfish_mouse_embryo' 
ref_tag   = 'embryo2_2'
out_dir   = file.path('output', tag)
obj_dir   = 'int_data/seqfish_mouse_embryo/embryo2_2'
dir.create(obj_dir)
obj_list  = list()
fig_dir   = 'figures/seqfish_mouse_embryo/embryo2_2'
dir.create(fig_dir)
fig_list = list()

methods     = list.files(out_dir)
names(methods) = methods
preds_f   = methods %>%
  map(~file.path(out_dir, .x, paste0(paste('preds', ref_tag, 'query_data', sep='_'), '.h5ad')))
imp_f = file.path('int_data/seqfish_mouse_embryo', 'gene_dt_embryo2_2.txt')
col_dt_f = file.path('int_data/seqfish_mouse_embryo', 'col_dt_embryo2_2.txt')

```





## Params
```{r setup_outputs, message=FALSE, cache=FALSE}
n_markers = 50
overwrite = FALSE
```

# Load inputs
```{r load_inputs, message=FALSE, cache=FALSE}
sce = preds_f[['sagenet']] %>% readH5AD 
cell_idx = which(sce$embryo != 'scRNAseq')
sce %<>% .[, cell_idx]
meta_dt = colData(sce) %>% as.data.table
if(overwrite | 
  !file.exists(file.path(obj_dir, 'map_2d.Rds')) | 
  !file.exists(file.path(obj_dir, 'meta_dt.Rds'))
){
  # Gene importanc

  preds_list = list()
  preds_list[['sagenet']] = anndata::read_h5ad(preds_f[['sagenet']])$obsm$dist_map 
  preds_list[['sagenet']]  %<>% as('sparseMatrix') %>% .[cell_idx, cell_idx]
  ## Novosparc predictions


gc()
}

imp = imp_f %>% fread %>% setnames('V1', 'GENE') %>%
  .[order(embryo2_2_entropy)] 
markers = imp$GENE[1:n_markers]
col_dt  = col_dt_f %>%  fread 
```

# Processing / calculations
## Distance matrices and embeddings
```{r calc_dists, message=FALSE, cache=FALSE}
if(overwrite | 
  !file.exists(file.path(obj_dir, 'map_2d.Rds')) | 
  !file.exists(file.path(obj_dir, 'meta_dt.Rds'))
){
  d = list()
  d_mtx = list()
  map_2d = list()

  # sagenet distances
  d[['sagenet']] = preds_list[['sagenet']] %>% as.dist
  d_mtx[['sagenet']] = d[['sagenet']] %>% as.matrix 
  rownames(d_mtx[['sagenet']]) = colnames(d_mtx[['sagenet']]) = colnames(sce)
  map_2d[['sagenet']] = Rtsne(d_mtx[['sagenet']], is_distance=T, perplexity=50)$Y
  rownames(map_2d[['sagenet']]) = colnames(sce)
  meta_dt[, `:=`(sagenet_1 = ..map_2d[['sagenet']][cell_id, 1], 
                sagenet_2 = ..map_2d[['sagenet']][cell_id, 2])]

  # obj_list[['d_mtx']] = append(d_mtx, d_mtx_true)
  obj_list[['map_2d']] = map_2d
  obj_list[['meta_dt']] = meta_dt
  gc()
}else{
  map_2d = readRDS(file.path(obj_dir, 'map_2d.Rds'))
  meta_dt = readRDS(file.path(obj_dir, 'meta_dt.Rds'))
}
```


## A: True regional labels + cell types (ref)
```{r A, fig.height=3, fig.width=10, message=FALSE, cache=FALSE}
p = plot_spatial(col_dt[, .(x, -y)], labels=factor(as.character(col_dt$cell_type), ordered=TRUE), label_cols=celltype_colours, hide_legend = TRUE, title = 'cell types') +coord_fixed()
p1 = plot_spatial(col_dt[, .(x, -y)], labels=factor(col_dt$leiden_0.01), hide_legend = TRUE, title='partitioning 1')+coord_fixed()
p2 = plot_spatial(col_dt[, .(x, -y)], labels=factor(col_dt$leiden_0.05), hide_legend = TRUE, title='partitioning 2')+coord_fixed()
p3 = plot_spatial(col_dt[, .(x, -y)], labels=factor(col_dt$leiden_0.1), hide_legend = TRUE, title='partitioning 3')+coord_fixed()
p4 = plot_spatial(col_dt[, .(x, -y)], labels=factor(col_dt$leiden_0.5), hide_legend = TRUE, title='partitioning 4')+coord_fixed()
p5 = plot_spatial(col_dt[, .(x, -y)], labels=factor(col_dt$leiden_1), hide_legend = TRUE, title='partitioning 5')+coord_fixed()

fig_list[['A']] = p + p1 + p2 + p3 + p4 + p5 + plot_layout(nrow=1) 
fig_list[['A']]
# plist
```

## B: UMAP plosts of expression space and the reconstructed space
```{r B, fig.height=5, fig.width=11, message=FALSE, cache=FALSE, eval=TRUE}
meta_dt[, ent :=  5 - (ent_embryo2_2_leiden_1 + ent_embryo2_2_leiden_0.1 + ent_embryo2_2_leiden_0.01 + ent_embryo2_2_leiden_0.5 + ent_embryo2_2_leiden_0.05) ]
meta_dt$ent = meta_dt$ent/(max(meta_dt$ent))

sagenet_p = plot_2d( 
    dim_df = meta_dt[, .(sagenet_1, sagenet_2)],
    labels = meta_dt$cell_type,
    label_cols = celltype_colours,
    hide_legend = TRUE,
    title = 'SageNet',
    sz = 1,
    alpha = meta_dt$ent,
) + labs(x = '', y = '')

sagenet_cont_p = plot_2d_cont( 
    dim_df = meta_dt[, .(sagenet_1, sagenet_2)],
    labels = meta_dt$ent,
    # label_cols = celltype_colours,
    hide_legend = FALSE,
    title = 'confidence',
    sz = 1,
    # alpha = meta_dt$ent,
) + labs(x = '', y = '', color='')



fig_list[['B']] = sagenet_p + sagenet_cont_p + plot_layout(nrow=1) 
fig_list[['B']]

```

## C: 
```{r C, fig.height=7, fig.width=11, message=FALSE, cache=FALSE}
tab_dt = col_dt[, .N, by = 'cell_type'] 
to_add = setdiff(meta_dt$cell_type, tab_dt$cell_type)
to_add_dt = data.table(cell_type = to_add, N = 0)
tab_dt = rbind(to_add_dt, tab_dt) %>% .[order(N), ]
celltype_ord = tab_dt$cell_type
p1 = tab_dt %>%
  ggplot +
  aes(x = factor(cell_type, ordered=TRUE, levels= celltype_ord), y = N, fill=cell_type) +
  geom_bar(stat="identity") +
  theme_bw() +
  scale_fill_manual(values=celltype_colours, na.value='gray') +
  labs(y='number of cells', x= '') +
  theme(legend.position='none') + 
  theme(axis.text.x=element_blank())

p2 = meta_dt %>%
  ggplot +
  aes(x = factor(cell_type, ordered=TRUE, levels= celltype_ord), y = ent, fill=cell_type) +
  geom_violin() +
  # + geom_jitter(height = 0, width = 0.1) +
  theme_bw() +
  scale_fill_manual(values=celltype_colours, na.value='gray') +
  labs(y='confidence score', x= '') +
  theme(legend.position='none') + 
  theme(axis.text.x=element_text(angle = 45, hjust = 1))

fig_list[['C']] = p1 + p2 + plot_layout(ncol = 1)
fig_list[['C']]
```


# Save objects & plots
```{r save, fig.height=12, fig.width=12, message=FALSE, cache=FALSE}
names(obj_list) %>% map(~saveRDS(obj_list[[.x]], file.path(obj_dir, paste0(.x, '.Rds'))))
names(fig_list) %>% map(~saveRDS(fig_list[[.x]], file.path(fig_dir, paste0(.x, '.Rds'))))
```
